### PR TITLE
Enhancements to fix blown stacks and make it easier to progress improvements

### PR DESCRIPTION
**Pull request by alistair-broomhead**

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -81,12 +81,33 @@ def swap_track(track_id):
     return [11, 14, 9, 16, 15, 10, 13, 12][track_id - 9]
 
 
+def tail_call_gen(func: typing.Callable[[...], typing.Generator]):
+    """
```
Comment (alistair-broomhead) [inline main.py:114]: I don't know if you've come across this syntax before but it's a docstring, which is essentially just a comment that's attached to a callable (i.e a class, a function or a method) to document what it does, this helps to differentiate versus comments about why a line of code is doing something a particular way

File: main.py [hunk 2]
```diff
@@ -1,12 +1,12 @@
-import sys  # used for setting recursion limit higher so program doesn't quit after a while
+import functools
+import typing
```
Comment (alistair-broomhead) [inline main.py:3]: This module just lets you add type hints for more stuff, which mostly just helps document an interface in a way that lets the IDE check your work

File: main.py [hunk 3]
```diff
@@ -81,12 +81,33 @@ def swap_track(track_id):
...
+    return facilitator
+
+
 # main generation function. uses given board and variables to generate tracks every game movement.
 # VERY condensed: it first looks if cars are on switches and does interaction shenanigans, then
 # generates tracks for every car BASED ON ITS VELOCITY. it then does checks for the generated tracks to determine
 # which ones are valid, and adds those as well as the generated car for each track to a list.
 # after every car is generated, it gets every possible track combination between every car, and calls the function
 # again with every new game movement.
+@tail_call_gen
```
Comment (alistair-broomhead) [inline main.py:139]: If you've not seen this syntax before, this is a decorator, and it's just syntactic sugar where

```python
@decorator
def function(*args):
    ...
```
is equivalent to

```python
def function(*args):
    ...
function = decorator(function)
```

File: main.py [hunk 4]
```diff
@@ -548,11 +569,13 @@ def generate_tracks(cars_to_use, board_to_use, ints_to_use, available_tracks, he
             continue
         if amt_placed_decoy == -1:
             return
-        generate_tracks(list(car_combos[combo_num]), board_to_pass, ints_to_pass,
-                        available_tracks - amt_placed_decoy, np.array(heatmaps), [list(solved[0]), list(solved[1])],
-                        stalled_to_pass, np.array(switch_queue), list(station_stalled),
-                        np.array(crashed_decoys), mvmts_since_solved, available_semaphores - placed_semaphores,
-                        heatmap_limits_pass)
+        yield (
```
Comment (alistair-broomhead) [inline main.py:601]: by using the yield statement you turn a function into a generator, which will run to the next yield statement every time it is iterated. This yields control to the caller, hence the statement name, and is also used to create coroutines, which are the basis of async and await, but that's a few steps of complexity beyond what's happening here

File: main.py [hunk 5]
```diff
@@ -1,12 +1,41 @@
...
 import itertools  # used for generating different board combinations
 
-sys.setrecursionlimit(100000)
+
+class PrinterProfile(cProfile.Profile):
+    """
+    Profiler context which always prints on context exit
+    if do_print is True
+    """
+    do_print = True
```
Comment (alistair-broomhead) [inline main.py:17]: I added this just to make it easier to disable printing

File: main.py [hunk 6]
```diff
@@ -1,12 +1,41 @@
...
+
+    def __init__(self, do_print=do_print, name=None, print_sort='cumtime'):
+        super().__init__()
+        self.print_sort = print_sort
+        self.name = name
+        self.do_print=do_print
+
+    def __enter__(self):
```
Comment (alistair-broomhead) [inline main.py:25]: These dunder-methods are what makes objects of this class usable as a context manager. when entering a `with` block then `__enter__` is called, and when leaving the block `__exit__` is called, whether it is through normal control low, or an exception

File: main.py [hunk 7]
```diff
@@ -548,97 +598,101 @@ def generate_tracks(cars_to_use, board_to_use, ints_to_use, available_tracks, he
...
-    elif len(lvl) == 5:
-        board, cars, maxTracks, interactions, decoys = lvl
-        if decoys is None:
-            decoys = []
-        semaphores = 0
-        ncars = []
-    elif len(lvl) == 6:
-        board, cars, maxTracks, interactions, decoys, semaphores = lvl
-        if decoys is None:
+for level_name in ["10-7"]:
```
Comment (alistair-broomhead) [inline main.py:612]: I did this just to make it so it was easier to differentiate the profile traces

File: main.py [hunk 8]
```diff
@@ -548,97 +598,101 @@ def generate_tracks(cars_to_use, board_to_use, ints_to_use, available_tracks, he
...
+    with PrinterProfile(name=level_name):
+        lvl = lc.levels[level_name]
+    # for key in lc.world7:
```
Comment (alistair-broomhead) [inline main.py:615]: Everything from here is just whitespace changes because everyhting is indented one level for the profiler context

**Comment by alistair-broomhead** [comment]
@FoxtrotOnce I've removed the commits which you didn't want

**Comment by FoxtrotOnce** [comment]
Thanks :)
The tail recursion should help a bit seeing it gave roughly a ~10% speed boost

**Comment by FoxtrotOnce** [comment]
@alistair-broomhead I was thinking of working on the project again this summer, since Afterburn announced a level editor is in the works. Would you like to work on it again? You were a huge help! I'm contacting you here since you left the Afterburn discord and you have direct messaging and friend requests turned off as well.

**Comment by alistair-broomhead** [comment]
@FoxtrotOnce would love to, sorry I left 'cause every time there was activity in the channels I followed it was spam, if the place is getting active again I'll happily come back if you pop an invite link here

**Pull request merged**